beacon-chain/sync/regular_sync.go

```diff
@@ -0,0 +1,321 @@
+func NewRegularSyncService(ctx context.Context, cfg *RegularSyncConfig) *RegularSync {
+	ctx, cancel := context.WithCancel(ctx)
+	return &RegularSync{
+		ctx:                   ctx,
+		cancel:                cancel,
+		p2p:                   cfg.P2P,
+		chainService:          cfg.ChainService,
+		db:                    cfg.BeaconDB,
+		attestationService:    cfg.AttestService,
+		blockAnnouncementFeed: new(event.Feed),
+		announceBlockBuf:      make(chan p2p.Message, cfg.BlockAnnounceBufferSize),
+		blockBuf:              make(chan p2p.Message, cfg.BlockBufferSize),
+		blockRequestBySlot:    make(chan p2p.Message, cfg.BlockRequestBufferSize),
+		attestationBuf:        make(chan p2p.Message, cfg.AttestationBufferSize),
+		chainHeadReqBuf:       make(chan p2p.Message, cfg.ChainHeadReqBufferSize),
+	}
+}
+
+// Start begins the block processing goroutine.
+func (rs *RegularSync) Start() {
+	go rs.run()
+}
+
+// ResumeSync resumes normal sync after initial sync is complete.
+func (rs *RegularSync) ResumeSync() {
+	go rs.run()
+}
+
+// Stop kills the block processing goroutine, but does not wait until the goroutine exits.
+func (rs *RegularSync) Stop() error {
+	log.Info("Stopping service")
+	rs.cancel()
+	return nil
+}
+
+// BlockAnnouncementFeed returns an event feed processes can subscribe to for
+// newly received, incoming p2p blocks.
+func (rs *RegularSync) BlockAnnouncementFeed() *event.Feed {
+	return rs.blockAnnouncementFeed
+}
+
+// run handles incoming block sync.
+func (rs *RegularSync) run() {
+	announceBlockSub := rs.p2p.Subscribe(&pb.BeaconBlockAnnounce{}, rs.announceBlockBuf)
+	blockSub := rs.p2p.Subscribe(&pb.BeaconBlockResponse{}, rs.blockBuf)
+	blockRequestSub := rs.p2p.Subscribe(&pb.BeaconBlockRequestBySlotNumber{}, rs.blockRequestBySlot)
+	attestationSub := rs.p2p.Subscribe(&pb.AggregatedAttestation{}, rs.attestationBuf)
+	chainHeadReqSub := rs.p2p.Subscribe(&pb.ChainHeadRequest{}, rs.chainHeadReqBuf)
+
+	defer announceBlockSub.Unsubscribe()
+	defer blockSub.Unsubscribe()
+	defer blockRequestSub.Unsubscribe()
+	defer chainHeadReqSub.Unsubscribe()
+	defer attestationSub.Unsubscribe()
+
+	for {
+		select {
+		case <-rs.ctx.Done():
+			log.Debug("Exiting goroutine")
+			return
+		case msg := <-rs.announceBlockBuf:
+			rs.receiveBlockAnnounce(msg)
+		case msg := <-rs.attestationBuf:
+			rs.receiveAttestation(msg)
+		case msg := <-rs.blockBuf:
+			rs.receiveBlock(msg)
+		case msg := <-rs.blockRequestBySlot:
+			rs.handleBlockRequestBySlot(msg)
+		case msg := <-rs.chainHeadReqBuf:
+			rs.handleChainHeadRequest(msg)
+		}
+	}
+}
+
+// receiveBlockHash accepts a block hash.
+// New hashes are forwarded to other peers in the network (unimplemented), and
+// the contents of the block are requested if the local chain doesn't have the block.
+func (rs *RegularSync) receiveBlockAnnounce(msg p2p.Message) {
+	ctx, receiveBlockSpan := trace.StartSpan(msg.Ctx, "receiveBlockHash")
```

Comment:
Nice, glad to see you are adding spans here!

Can we use the span name format of struct_method?
For this: `RegularSync_receiveBlockHash`, it will be easier to determine what is going on from search results in jaeger.

Reply:
ah no actually , this was just copy and pasted code. But yeah I can change the format of the message. Should we start adding spans to all out p2p subscriptions in our services?